### ISSUE TITLE
Validation loader flat loss

### ISSUE THREAD

**Issue by Dewald928** [davidtvs/pytorch-lr-finder#59]
I copied your example notebook to colab and ran the code without changing anything. But the validation loss I get goes flat, which is clearly a mistake when compared to your example. I also experienced this with my other networks which do the same, the loss just goes flat.

You can see my results from colab and your example in the figures below.

| Colab | Your Example notebook |
|---|---|
|![colab](https://user-images.githubusercontent.com/14140515/89273404-c97e4200-d63f-11ea-970a-b62d28ce46cd.png)  | ![notebook](https://user-images.githubusercontent.com/14140515/89273418-cd11c900-d63f-11ea-923e-83f10acfc7b1.png) |

**Comment by NaleRaphael** [comment]
@Dewald928 Thanks for reporting this.

After a quick run on that notebook ("examples/lrfinder_mnist.ipynb"), I found that running `range_test()` with `val_loader` is as quick as without `val_loader`, which is absolutely abnormal. I'll keep investigating it.

**Comment by NaleRaphael** [comment]
OK, I figure out the reason why `range_test()` is running as quickly as it runs without a `val_loader`.

In `range_test()`, the following loop works normally at the first iteration:
```python
# @LRFinder.range_test()
# ...
for iteration in tqdm(range(num_iter)):
    # Train on batch and retrieve loss
    loss = self._train_batch(
        train_iter,
        accumulation_steps,
        non_blocking_transfer=non_blocking_transfer,
    )
    if val_loader:
        loss = self._validate(
            val_iter, non_blocking_transfer=non_blocking_transfer
        )
# ...
```

However, `val_iter._iterator` has run out of values after that iteration and won't be reset in the following execution. Hence that `self._validate()` won't do anything and just return the default output: 0.0 (known as `running_loss` in that method). Therefore, `loss` returned by `self._train_batch()` is overwritten by 0.0, and it will be re-calculated by the following code:
```python
# Track the best loss and smooth it if smooth_f is specified
if iteration == 0:
    self.best_loss = loss
else:
    if smooth_f > 0:
        loss = smooth_f * loss + (1 - smooth_f) * self.history["loss"][-1]
    if loss < self.best_loss:
        self.best_loss = loss
```

And that's why the lr-loss curve goes flat like the result you provided.

I'll make a patch for it later.
